src/file.rs

```diff
@@ -65,106 +74,134 @@ pub fn madvise_random(map: &mut memmap2::MmapMut) {
-fn mmap(file: &std::fs::File, len: usize) -> Result<memmap2::MmapMut> {
+fn mmap(file: &std::fs::File, len: usize, max_size: Option<usize>) -> Result<memmap2::MmapMut> {
 	#[cfg(not(test))]
 	const RESERVE_ADDRESS_SPACE: usize = 1024 * 1024 * 1024; // 1 Gb
 														 // Use a different value for tests to work around docker limits on the test machine.
 	#[cfg(test)]
 	const RESERVE_ADDRESS_SPACE: usize = 64 * 1024 * 1024; // 64 Mb
 
 	let map_len = len + RESERVE_ADDRESS_SPACE;
+	let map_len = max_size.map(|m| std::cmp::min(map_len, m)).unwrap_or(map_len);
 	let mut map = try_io!(unsafe { memmap2::MmapOptions::new().len(map_len).map_mut(file) });
 	madvise_random(&mut map);
 	Ok(map)
 }
 
 #[cfg(windows)]
-fn mmap(file: &std::fs::File, _len: usize) -> Result<memmap2::MmapMut> {
+fn mmap(file: &std::fs::File, _len: usize, _max_size: Option<usize>) -> Result<memmap2::MmapMut> {
 	Ok(try_io!(unsafe { memmap2::MmapOptions::new().map_mut(file) }))
 }
 
 const GROW_SIZE_BYTES: u64 = 256 * 1024;
 
 #[derive(Debug)]
 pub struct TableFile {
-	pub map: RwLock<Option<(memmap2::MmapMut, std::fs::File)>>,
-	pub path: std::path::PathBuf,
+	pub maps: RwLock<Vec<(memmap2::MmapMut, std::fs::File)>>,
+	pub path_base: std::path::PathBuf,
 	pub capacity: AtomicU64,
 	pub id: TableId,
+	pub max_size: Option<usize>,
 }
 
 impl TableFile {
-	pub fn open(filepath: std::path::PathBuf, entry_size: u16, id: TableId) -> Result<Self> {
-		let mut capacity = 0u64;
-		let map = if std::fs::metadata(&filepath).is_ok() {
-			let file = try_io!(std::fs::OpenOptions::new()
-				.read(true)
-				.write(true)
-				.disable_read_ahead()
-				.open(filepath.as_path()));
-			try_io!(disable_read_ahead(&file));
-			let len = try_io!(file.metadata()).len();
-			if len == 0 {
-				// Preallocate.
-				capacity += GROW_SIZE_BYTES / entry_size as u64;
-				try_io!(file.set_len(GROW_SIZE_BYTES));
-			} else {
-				capacity = len / entry_size as u64;
-			}
-			let map = mmap(&file, len as usize)?;
-			Some((map, file))
+	pub fn open(
+		path_base: std::path::PathBuf,
+		entry_size: u16,
+		id: TableId,
+		max_file_size: Option<usize>,
+	) -> Result<Self> {
+		let max_size = if let Some(m) = max_file_size {
+			let m = m * 1024 * 1024;
+			Some((m / entry_size as usize) * entry_size as usize)
 		} else {
 			None
 		};
+		let mut capacity = 0u64;
+		let mut maps = Vec::new();
+		for i in 0.. {
+			let mut filepath = path_base.clone();
+			filepath.push(id.file_name(max_file_size.is_some().then(|| i)));
+			if std::fs::metadata(&filepath).is_ok() {
+				let file = try_io!(std::fs::OpenOptions::new()
+					.read(true)
+					.write(true)
+					.disable_read_ahead()
+					.open(filepath.as_path()));
+				try_io!(disable_read_ahead(&file));
+				let len = try_io!(file.metadata()).len();
+				if len == 0 {
+					// Preallocate.
+					capacity += GROW_SIZE_BYTES / entry_size as u64;
+					try_io!(file.set_len(GROW_SIZE_BYTES));
+				} else {
+					capacity = len / entry_size as u64;
```

Comment:
Think this should be `+=`.  Though it doesn't cause any obvious problems as if the value is incorrect then grow is called and the correct capacity is calculated there.

Reply:
Thanks, sure `+=` is right here.